File: typescript/packages/core/src/core/gdocs/unlink.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GDocsAccessor } from '../../accessor/gdocs.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { stripSlash } from '../../util/slash.ts'
import { deleteFile } from '../google/drive.ts'
import { readdir as coreReaddir } from './readdir.ts'

const VIRTUAL_DIRS = new Set(['', 'owned', 'shared'])

function enoent(p: string): Error & { code: string } {
  const e = new Error(`ENOENT: ${p}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

function eisdir(p: string): Error & { code: string } {
  const e = new Error(`EISDIR: ${p}`) as Error & { code: string }
  e.code = 'EISDIR'
  return e
}

export async function unlink(
  accessor: GDocsAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<void> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (VIRTUAL_DIRS.has(key)) throw eisdir(path.original)
  if (index === undefined) throw enoent(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`
  let result = await index.get(virtualKey)
  if (result.entry === undefined || result.entry === null) {
    const parentVirtual = virtualKey.includes('/')
      ? virtualKey.slice(0, virtualKey.lastIndexOf('/')) || '/'
      : '/'
    try {
      await coreReaddir(
        accessor,
        new PathSpec({
          original: parentVirtual,
          directory: parentVirtual,
          resolved: false,
          prefix,
        }),
        index,
      )
    } catch {
      // parent listing failed — fall through to not-found
    }
    result = await index.get(virtualKey)
  }
  if (result.entry === undefined || result.entry === null) throw enoent(path.original)
  await deleteFile(accessor.tokenManager, result.entry.id)
  const parentDir = virtualKey.includes('/')
    ? virtualKey.slice(0, virtualKey.lastIndexOf('/')) || '/'
    : '/'
  await index.invalidateDir(parentDir)
}

File: python/mirage/core/gslides/unlink.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.gslides import GSlidesAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.google.drive import delete_file
from mirage.core.gslides.readdir import readdir
from mirage.types import PathSpec
from mirage.utils.errors import enoent

VIRTUAL_DIRS = {"", "owned", "shared"}


async def unlink(
    accessor: GSlidesAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> None:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    prefix = path.prefix
    raw = path.original
    stripped = raw[len(prefix):] if prefix and raw.startswith(prefix) else raw
    key = stripped.strip("/")
    if key in VIRTUAL_DIRS:
        raise IsADirectoryError(raw)
    if index is None:
        raise enoent(path)
    virtual_key = prefix + "/" + key if prefix else "/" + key
    result = await index.get(virtual_key)
    if result.entry is None:
        parent = "/" + "/".join(key.split("/")[:-1])
        parent_path = PathSpec.from_str_path(prefix + parent, prefix=prefix)
        await readdir(accessor, parent_path, index)
        result = await index.get(virtual_key)
    if result.entry is None:
        raise enoent(path)
    await delete_file(accessor.token_manager, result.entry.id)
    parent_dir = "/".join(virtual_key.rsplit("/", 1)[:-1]) or "/"
    await index.invalidate_dir(parent_dir)
